src/psyclone/psyir/symbols/symbol_table.py

```diff
@@ -1396,6 +1396,35 @@ def resolve_imports(self, container_symbols=None, symbol_target=None):
                                             symbol, symbol_target):
                     continue
 
+                # Determine if there is an Unresolved Symbol in a
+                # descendent symbol table that matches the name of the
+                # symbol we are importing and if so, move it to this
+                # symbol table if a symbol with the same name does not
+                # already exist in this symbol table.
+
+                # There are potential issues with this approach and
+                # with the routine in general which are captured in
+                # issue #2331. Issue #2271 may also help/fix some or
+                # all of the problems too.
+
+                # Import here to avoid circular dependencies
+                # pylint: disable=import-outside-toplevel
+                from psyclone.psyir.nodes import ScopingNode
+                for scoping_node in self.node.walk(ScopingNode):
+                    symbol_table = scoping_node.symbol_table
+                    if symbol.name in symbol_table:
+                        test_symbol = symbol_table.lookup(symbol.name)
+                        # pylint: disable=unidiomatic-typecheck
+                        if (type(test_symbol) == Symbol
+                            and test_symbol.is_unresolved
+                                and test_symbol.name not in self):
+                            # No wildcard imports in this symbol table
+                            if not [sym for sym in
+                                    symbol_table.containersymbols if
+                                    sym.wildcard_import]:
+                                symbol_table.remove(test_symbol)
+                                self.add(test_symbol)
```

Comment:
There is symbol_table.swap(old_symbol, new_symbol) to do this, but the problem is that this will hit issue #898 and will not update that actual use of the symbols inside the tree, which is quite dangerous (e.g renames will not work, next_available_name will not see it, ...).

Reply:
Seeing as we're inlining, does that make it less dangerous? Are you asking me to change the implementation to use swap. Sorry, it is not clear (to me).

Reply:
As we discussed, the problem is that we have the symbol_table AND the references nodes storing the Symbol object.

When we remove a symbol from a symbol_table and then add it higher up, everything is fine. The reference nodes ref.symbol still exists in an ancestor symbol table.

However, when we remove the symbol, and expect a higher up symbol_table to already have a symbol with the same name, we leave the ref.symbol inside the tree pointing to an object that has been removed. So at least we need to do:
```
else: 
  for ref in symbol_table.node.walk(Reference):
     if SymbolTable._has_same_name(ref.symbol, test_symbol):
         ref.symbol = test_symbol
```
A quick way to see if this worked is by taking the test case suggested in the comment below and after resolving (and hence moving the symbols to the module) and then do a `symbol_table.rename_symbol(symbol_table.lookup("some_var"), "other_var")` and see if all references have been updated.